src/util/permissions.ts

```diff
@@ -1,251 +1,260 @@
+export async function hasRolePermissions(
   memberID: string,
+  guildID: string,
   permissions: Permission[],
 ) {
-  const channel = await cacheHandlers.get("channels", channelID);
-  if (!channel) return false;
-  if (!channel.guildID) return true;
+  // First we need the role permission bits this member has
+  const basePermissions = await calculateBasePermissions(memberID, guildID);
+  // Second use the validatePermissions function to check if the member has every permission
+  return validatePermissions(basePermissions, permissions);
+}
 
-  const guild = await cacheHandlers.get("guilds", channel.guildID);
-  if (!guild) return false;
+/** Checks if the bot has these permissions in the given guild */
+export function botHasRolePermissions(
+  guildID: string,
+  permissions: Permission[],
+) {
+  // Since Bot is a normal member we can use the hasRolePermissions() function
+  return hasRolePermissions(guildID, botID, permissions);
+}
 
-  if (guild.ownerID === memberID) return true;
-  if (
-    await memberIDHasPermission(memberID, guild.id, ["ADMINISTRATOR"])
-  ) {
-    return true;
-  }
-  const member = (await cacheHandlers.get("members", memberID))?.guilds.get(
-    guild.id,
+/** Returns the permissions that are not in the given permissionBits */
+export function missingPermissions(
+  permissionBits: string,
+  permissions: Permission[],
+) {
+  if (BigInt(permissionBits) & 8n) return [];
+
+  return permissions.filter((permission) =>
+    !(BigInt(permissionBits) & BigInt(permission))
   );
-  if (!member) return false;
-
-  let memberOverwrite: RawOverwrite | undefined;
-  let everyoneOverwrite: RawOverwrite | undefined;
-  const rolesOverwrites: RawOverwrite[] = [];
-
-  for (const overwrite of channel.permissionOverwrites || []) {
-    // If the overwrite on this channel is specific to this member
-    if (overwrite.id === memberID) memberOverwrite = overwrite;
-    // If it is the everyone role overwrite
-    if (overwrite.id === guild.id) everyoneOverwrite = overwrite;
-    // If it is one of the roles the member has
-    if (member.roles.includes(overwrite.id)) rolesOverwrites.push(overwrite);
-  }
+}
 
-  const allowedPermissions = new Set<Permission>();
-
-  // Member perms override everything so we must check them first
-  if (memberOverwrite) {
-    const allowBits = memberOverwrite.allow;
-    const denyBits = memberOverwrite.deny;
-    for (const perm of permissions) {
-      // One of the necessary permissions is denied. Since this is main permission we can cancel if its denied.
-      if (BigInt(denyBits) & BigInt(Permissions[perm])) return false;
-      // Already allowed perm
-      if (allowedPermissions.has(perm)) continue;
-
-      // This perm is allowed so we save it
-      if (BigInt(allowBits) & BigInt(Permissions[perm])) {
-        allowedPermissions.add(perm);
-      }
-    }
+/** Throws an error if this member has not all of the given permissions */
+export async function requireGuildPermissions(
```

Comment:
```suggestion
export async function requirePermissions(
```

Reply:
Not sure about that. There is `requireChannelPermission` too.

Reply:
The name does not make sense, are there permissions for DM channels?

Reply:
`requireGuildPermissions` does only calculate guild ones.
`requireChannelPermission` does both if its a DM it returns Administrator.